src/process_utility.c

```diff
@@ -1263,43 +1430,112 @@ process_index_end(Node *parsetree, CollectedCommand *cmd)
 	hcache = ts_hypertable_cache_pin();
 	ht = ts_hypertable_cache_get_entry_rv(hcache, stmt->relation);
 
-	if (NULL != ht)
+	if (NULL == ht)
 	{
-		CreateIndexInfo info = {
-			.stmt = stmt,
-		};
+		ts_cache_release(hcache);
+		return false;
+	}
 
-		switch (cmd->type)
+	switch (cmd->type)
+	{
+		case SCT_Simple:
+			info.obj = cmd->d.simple.address;
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("hypertables do not support this operation: "
+							"parsetree %s, type %d",
+							nodeToString(parsetree),
+							cmd->type)));
+			break;
+	}
+
+	if (info.extended_options.multitransaction)
+	{
+		Relation index = relation_open(info.obj.objectId, AccessShareLock);
+		LockRelId index_lock_relid = index->rd_lockInfo.lockRelId;
+
+		relation_close(index, NoLock);
+
+		/*
+		 * Lock the index for the remainder of the command. Since we're using
+		 * multiple transactions for index creation, a regular
+		 * tranaction-level lock won't prevent the index from being
+		 * concurrently ALTERed or DELETEed. Instead, we grab a session level
+		 * lock on the index, which we'll release when the command is
+		 * finished. (This is the same strategy postgres uses in CREATE INDEX
+		 * CONCURRENTLY)
+		 */
+		LockRelationIdForSession(&index_lock_relid, AccessShareLock);
+
+		/*
+		 * mark the hypertable's index as invalid until all the chunk indexes
+		 * are created. This allows us to determine if the CREATE INDEX
+		 * completed successfully or  not
+		 */
+		ts_indexing_mark_as_invalid(info.obj.objectId);
+		CacheInvalidateRelcacheByRelid(ht->main_table_relid);
+		CacheInvalidateRelcacheByRelid(info.obj.objectId);
+
+		ts_cache_release(hcache);
+
+		info.mctx = CurrentMemoryContext;
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		/*
+		 * foreach_chunk needs an open tranaction, it will be closed in the
+		 * first chunk
+		 */
+		StartTransactionCommand();
+		MemoryContextSwitchTo(info.mctx);
+
+		LockRelationOid(ht->main_table_relid, AccessShareLock);
+
+		foreach_chunk_relid(RangeVarGetRelid(stmt->relation, NoLock, false),
+							process_index_chunk_multitransaction,
+							&info);
+
+		/*
+		 * we need a new transaction, b/c if we had no chunks, we will have
+		 * already updated the index validity tuple
+		 */
+		CommitTransactionCommand();
+
+		StartTransactionCommand();
+		MemoryContextSwitchTo(info.mctx);
+
+		if (info.extended_options.max_chunks < 0)
```

Comment:
wait a minute. Not so fast. What happens if we created a chunk after foreach_chunk started? Would it create the index when the chunk is created? would that index be marked invalid then because it's marked invalid on the hypertable? Also isn't max_chunks a debug only thing?

Reply:
yes, the regular chunk creation code creates the chunk; the root index is created fine, just marked invalid, and our chunk creation code does not care.
max_chunks is only a testing thing, currently it handles that by hardcoding to `-1` in in all other cases.